crypto/dilithium/README.md

```diff
@@ -40,4 +43,34 @@ EVP_DigestSign(md_ctx.get(), signature, &signature_len, msg, msg_len);
 
 // Verify
 EVP_DigestVerify(md_ctx.get(), signature, signature_len, msg, msg_len);
-```
\ No newline at end of file
+```
+
+You can also sign Dilithium3 signatures using a `pem` private key file and the `bssl sign` tool. For example, to sign raw standard input you can use `echo "foo" | ./bssl sign -key private.pem`.
+
+## Benchmarks
+Benchmarks for KeyGen, Sign, and Verify functionality are available within the `bssl speed` tool, and can be explicitly filtered by running `./tool/bssl speed -filter Dilithium3`.
+
+Example output:
+```
+Did 4830 Dilithium3 keygen operations in 1014176us (4762.5 ops/sec)
+Did 2688 Dilithium3 signing operations in 1062613us (2529.6 ops/sec)
+Did 5124 Dilithium3 verify operations in 1014016us (5053.2 ops/sec)
+```
+
+Open Quantum Safe (OQS) provides benchmarks for the liboqs implementation of Dilithium based upon the same reference implementation. Full OQS benchmarks are available at [OQS benchmarking](https://openquantumsafe.org/benchmarking/visualization/speed_sig.html). For comparison we include the OQS Dilithium3 benchmarks from 2022-12-07.
+
+``` 
++-------------------------------+-----------+-----------------+---------+--------------+----------+----------------+
+|           Algorithm           | keypair/s | keypair(cycles) | sign/s  | sign(cycles) | verify/s | verify(cycles) |
++-------------------------------+-----------+-----------------+---------+--------------+----------+----------------+
+| Dilithium3 AWS-LC(x86_64-ref) |    4762.5 |          524934 |  2529.6 |       988298 |   5053.2 |         494736 |
+| Dilithium3 OQS (x86_64-ref)   |    5315.0 |          470266 |  1358.3 |      1840581 |   5690.7 |         439257 |
++-------------------------------+-----------+-----------------+---------+--------------+----------+----------------+
```

Comment:
Historically we haven't published performance numbers in our readme's @dkostic what do you think? This isn't offering a performance guarantee, but is this data valuable on it's own? Should we include the gen/sign/verify time of all algorithms?

Reply:
While the numbers we present are not a guarantee, it would be bad customer experience if the numbers didn't match the reality, so we'd have to updated the numbers every time there is a change. But not only changes of the core primitive can affect the performance. For example, Dilithium3 goes through layers of EVP signature implementation so any change in those may impact the performance. Granted, I don't expect this to happen often, but because it can happen we'd have to check and update the numbers accordingly. Then the next set of problems arises -- we need to ensure that every time we measure the performance in exactly the same conditions (CPU, OS, compiler, build flags, etc.), that the measured performance is not an outlier, etc. While this might be doable for a single algorithm, I'm concerned about scalability and maintainability. I think a much better solution would be to make the data from the internal performance dashboards publicly available in some way.

The comparison with libOQS for Dilithium is more problematic because:
 - we can't ensure that the numbers libOQS presents were obtained in the same conditions as our numbers (CPU is not the only variable in the system, especially when we are dealing with C code),
 - the code flow in libOQS (OpenSSL) and AWS-LC is not the same which means we are not comparing apples to apples. For example, the numbers Jake added show that signing is almost twice slower in AWS-LC, while theoretically they should be the same because we use exactly the same core implementation of Dilithium. How can we explain this? And why is there no such difference in other operations?
We've already been dealing with these problems in the benchmarking tool and we have a solution, so again, I'd rather add Dilithium to the general benchmarking tool. And if we really want to compare with libOQS we should add it to the tool as well.

A bit off-topic: @jakemas how did you get the number of cycles?

Reply:
Agree with Dusan here.
> We've already been dealing with these problems in the benchmarking tool and we have a solution, so again, I'd rather add Dilithium to the general benchmarking tool. And if we really want to compare with libOQS we should add it to the tool as well.
What tool is this? bssl speed? If so, this PR is doing exactly that - adding dilithium to the general benchmarking tool. I would just remove OQS and the table to be consistent with other benchmarks.